File: server-csharp/OpenRSC.Server/Actions/WalkToAction.cs
using OpenRSC.Server.Entities;
using OpenRSC.Server.Models;

namespace OpenRSC.Server.Actions;

/// <summary>
/// Base class for all walk-to actions. Players walk to a target location
/// and then execute an action when they arrive.
/// </summary>
public abstract class WalkToAction
{
    private volatile bool _executed;
    private int _retryAttempts;
    private long _lastAttemptTick;

    /// <summary>
    /// The player performing this action.
    /// </summary>
    public Player Player { get; }

    /// <summary>
    /// Target location for this action.
    /// </summary>
    public Point Location { get; }

    /// <summary>
    /// Maximum number of retry attempts.
    /// </summary>
    public int MaxRetries { get; set; }

    /// <summary>
    /// Whether retry is enabled for this action.
    /// </summary>
    public bool RetryEnabled { get; set; }

    protected WalkToAction(Player player, Point location)
    {
        Player = player;
        Location = location;
        _executed = false;
        _retryAttempts = 0;
        _lastAttemptTick = 0;
        MaxRetries = player.ActionRetryConfig.MaxRetries;
        RetryEnabled = player.ActionRetryConfig.Enabled;
    }

    /// <summary>
    /// Executes the action if conditions are met.
    /// </summary>
    public void Execute()
    {
        ExecuteInternal();
        FinishExecution();
    }

    /// <summary>
    /// Marks the action as executed and records it.
    /// </summary>
    protected void FinishExecution()
    {
        IsExecuted = true;
        Player.SetLastExecutedWalkToAction(this);
    }

    /// <summary>
    /// Checks if this action should execute on this tick.
    /// </summary>
    public bool ShouldExecute() => !IsExecuted && ShouldExecuteInternal();

    /// <summary>
    /// Override to implement the actual action logic.
    /// </summary>
    protected abstract void ExecuteInternal();

    /// <summary>
    /// Override to implement the condition check for execution.
    /// </summary>
    protected abstract bool ShouldExecuteInternal();

    /// <summary>
    /// Whether this action has been executed.
    /// </summary>
    public bool IsExecuted
    {
        get => _executed;
        private set => _executed = value;
    }

    /// <summary>
    /// Whether this is a PvP attack action.
    /// </summary>
    public virtual bool IsPvPAttack => false;

    #region Retry Mechanism

    /// <summary>
    /// Called when ShouldExecute returns false. Increments attempt counter.
    /// </summary>
    /// <param name="currentTick">The current game tick</param>
    /// <returns>True if action should be retried, false if max retries exceeded</returns>
    public bool OnAttemptFailed(long currentTick)
    {
        if (!RetryEnabled)
            return false;

        _lastAttemptTick = currentTick;
        _retryAttempts++;
        return _retryAttempts < MaxRetries;
    }

    /// <summary>
    /// Checks if this action has exceeded its retry limit.
    /// </summary>
    public bool HasExceededRetryLimit => RetryEnabled && _retryAttempts >= MaxRetries;

    /// <summary>
    /// Gets the number of retry attempts made.
    /// </summary>
    public int RetryAttempts => _retryAttempts;

    /// <summary>
    /// Gets the last tick when an attempt was made.
    /// </summary>
    public long LastAttemptTick => _lastAttemptTick;

    /// <summary>
    /// Resets the retry counter.
    /// </summary>
    public void ResetRetryAttempts() => _retryAttempts = 0;

    /// <summary>
    /// Gets the failure message when max retries are exceeded.
    /// Override in subclasses for custom messages.
    /// </summary>
    public virtual string FailureMessage => "You are unable to reach your target.";

    #endregion
}

File: server/src/com/openrsc/server/model/action/WalkToAction.java
package com.openrsc.server.model.action;

import com.openrsc.server.model.Point;
import com.openrsc.server.model.entity.player.Player;

public abstract class WalkToAction {

	private Player player;
	private Point location;
	private volatile boolean executed;

	// Retry mechanism fields
	private int retryAttempts;
	private int maxRetries;
	private long lastAttemptTick;
	private boolean retryEnabled;

	public WalkToAction(final Player player, final Point location) {
		this.player = player;
		this.location = location;
		this.executed = false;
		this.retryAttempts = 0;
		this.maxRetries = player.getConfig().ACTION_MAX_RETRIES;
		this.lastAttemptTick = 0;
		this.retryEnabled = player.getConfig().ACTION_RETRY_ENABLED;
	}

	public void execute() {
		executeInternal();
		finishExecution();
	}

	public void finishExecution() {
		setExecuted(true);
		player.setLastExecutedWalkToAction(this);
	}

	public final boolean shouldExecute() {
		return !isExecuted() && shouldExecuteInternal();
	}

	protected abstract void executeInternal();

	protected abstract boolean shouldExecuteInternal();

	public Player getPlayer() {
		return player;
	}

	public Point getLocation() {
		return location;
	}

	public synchronized boolean isExecuted() {
		return executed;
	}

	protected synchronized void setExecuted(boolean executed) {
		this.executed = executed;
	}

	public boolean isPvPAttack() {
		return false;
	}

	// Retry mechanism methods

	/**
	 * Called when shouldExecute returns false. Increments attempt counter.
	 * @param currentTick The current game tick
	 * @return true if action should be retried, false if max retries exceeded
	 */
	public boolean onAttemptFailed(final long currentTick) {
		if (!retryEnabled) {
			return false;
		}
		this.lastAttemptTick = currentTick;
		this.retryAttempts++;
		return retryAttempts < maxRetries;
	}

	/**
	 * Checks if this action has exceeded its retry limit
	 * @return true if max retries have been exceeded
	 */
	public boolean hasExceededRetryLimit() {
		return retryEnabled && retryAttempts >= maxRetries;
	}

	/**
	 * Checks if retry is enabled for this action
	 * @return true if retry is enabled
	 */
	public boolean isRetryEnabled() {
		return retryEnabled;
	}

	/**
	 * Sets whether retry is enabled for this action
	 * @param retryEnabled true to enable retry
	 */
	public void setRetryEnabled(boolean retryEnabled) {
		this.retryEnabled = retryEnabled;
	}

	/**
	 * Gets the number of retry attempts made
	 * @return the number of retry attempts
	 */
	public int getRetryAttempts() {
		return retryAttempts;
	}

	/**
	 * Gets the maximum number of retries allowed
	 * @return the maximum retries
	 */
	public int getMaxRetries() {
		return maxRetries;
	}

	/**
	 * Sets the maximum number of retries allowed
	 * @param maxRetries the maximum retries
	 */
	public void setMaxRetries(int maxRetries) {
		this.maxRetries = maxRetries;
	}

	/**
	 * Resets the retry counter (useful if conditions improve)
	 */
	public void resetRetryAttempts() {
		this.retryAttempts = 0;
	}

	/**
	 * Gets the last tick when an attempt was made
	 * @return the last attempt tick
	 */
	public long getLastAttemptTick() {
		return lastAttemptTick;
	}

	/**
	 * Override this method to provide a custom failure message when max retries are exceeded.
	 * @return the failure message to display to the player
	 */
	public String getFailureMessage() {
		return "You are unable to reach your target.";
	}
}
